aether-kernel/aether/kernel/api/project_artefacts.py

```diff
@@ -184,7 +185,12 @@ def upsert_project_artefacts(
             }
 
         # check for the mapping set
-        mappingset_id = raw_mapping.get('mappingset', raw_mapping.get('id'))
+        mappingset_id = raw_mapping.get('mappingset')
+        if not mappingset_id:
+            if mappingset:
```

Comment:
why are you using a global mappingset?

Reply:
To allow a one-many pairing for mappings to a mappingset in one call.

Reply:
Can you write and example of that case? I don't understand why using a global mappingset helps in that case...

Reply:
Take in mind that with your approach you are always using the LAST mappingset, you need to indicate the specific mappingset for the mapping, otherwise you create a new mappingset that shares the mapping id.

When you want to create a "group" (avoiding set here) mappings assigned to a certain mappingset you create the mappingset providing its id and the mappings using that id.

```json
{
  "mappingsets": [
    {"id": 1, "name": "1"}, 
    {"id": 2, "name": "2"}, 
    {"name": "with-a-random-id"}
  ],
  "mappings": [
    {"name": "m-1-1", "mappingset": 1},
    {"name": "m-1-2", "mappingset": 1},
    {"name": "m-2-1", "mappingset": 2},
    {"name": "m-2-2", "mappingset": 2},
    {"name": "with-a-new-mappingset"},
  ]
}
```

Reply:
In a case where you can't know the assigned `mappingset.id` to set the `mapping.mappingset` field before making the call.

Assuming I have the payload: 
```
{
   mappingsets: [
       {name: 'mappingset-1'}
   ],
   mappings: [
     {name: 'mapping-1'},
     {name: 'mapping-2'}
   ]
}
```
Note: without any `ids`.
I will assume the both mappings should belong to the same mappingset. With the previous implementation you will end up with 3 mappingsets and the mappings each linked to one of the two extra mappingsets created.

Reply:
Moreover, I would like to think of the concept of posting 'artefacts' in the context of a `pipeline` = 1 `mappinset`. Be it from the `ui` or `gather` its always in the context of creating a (1) pipeline within a project

Reply:
My problem now is that you are breaking this method and when Gather propagates the xForms from ODK to Kernel it's creating more than ONE mappingset.... and the side effect can be a nightmare. :disappointed: